libphal/phal_sbe.C

```diff
@@ -15,6 +20,80 @@ using namespace openpower::phal;
 using namespace openpower::phal::utils::pdbg;
 using namespace openpower::phal::pdbg;
 
+/**
+ * @brief helper function to perform HRESET on secondary processor
+ *        if sbe is in HALT state.
+ * @param[in] proc processor target to operate on
+ *
+ * Exceptions: SbeError/PdbgError with failure reason code
+ */
+void sbeHaltStateRecovery(struct pdbg_target *proc)
+{
+	// Skip for primary processor.
+	if (isPrimaryProc(proc)) {
+		return;
+	}
+	// Required to probe processor associated fsi to run hardware procedures
+	char path[16];
+	struct pdbg_target *fsi;
+	sprintf(path, "/proc%d/fsi", pdbg_target_index(proc));
+	fsi = pdbg_target_from_path(NULL, path);
+	if (fsi != nullptr) {
+		log(level::ERROR, "Failed to get fsi target on (%s)",
+		    pdbg_target_path(proc));
+		throw pdbgError_t(exception::PDBG_TARGET_NOT_OPERATIONAL);
+	}
+
+	if (pdbg_target_probe(fsi) != PDBG_TARGET_ENABLED) {
+		log(level::ERROR, "fsi(%s) probe: fail to enable",
+		    pdbg_target_path(fsi));
+		throw pdbgError_t(exception::PDBG_TARGET_NOT_OPERATIONAL);
+	}
+
+	// get SBE current state based on sbe message register
+	sbeMsgReg_t sbeReg;
+	fapi2::ReturnCode fapiRC;
+	fapiRC = p10_get_sbe_msg_register(proc, sbeReg);
+	if (fapiRC != fapi2::FAPI2_RC_SUCCESS) {
+		log(level::ERROR,
+		    "Failed: checking SBE state in Halt (%s) fapiRC=%x",
+		    pdbg_target_path(proc), fapiRC);
+		throw sbeError_t(exception::SBE_STATE_READ_FAIL);
+	}
+
+	// Check SBE state is in halt
+	if (sbeReg.currState != SBE_STATE_HALT) {
+		// No special actions required
+		return;
+	}
+
+	log(level::INFO, "SBE(%s) HRESET Requested", pdbg_target_path(proc));
+
+	// SBE is in Halt state, call p10_sbe_reset hwp and re-check state
+	fapiRC = p10_sbe_hreset(proc, true);
+	if (fapiRC != fapi2::FAPI2_RC_SUCCESS) {
+		log(level::ERROR, "failed hreset hwp(%s), fapiRC=%x",
+		    pdbg_target_path(proc), fapiRC);
+		throw sbeError_t(exception::HWP_EXECUTION_FAILED);
+	}
+
+	// wait for 5 seconds for the change in SBE state
+	sleep(5);
+
+	// Check sbe current state using message register
+	fapiRC = p10_get_sbe_msg_register(proc, sbeReg);
+	if (fapiRC != fapi2::FAPI2_RC_SUCCESS) {
+		log(level::ERROR,
+		    "Failed: Checking SBE state procedure (%s) fapiRC=%x",
+		    pdbg_target_path(proc), fapiRC);
+		throw sbeError_t(exception::SBE_STATE_READ_FAIL);
+	}
+	if (sbeReg.currState == SBE_STATE_RUNTIME) {
+		log(level::INFO, "SBE (%s), recoverd from halt state",
+		    pdbg_target_path(proc));
+	}
```

Comment:
do we need to throw error if sbe does not reachruntime, is there any purpose to continue

Reply:
Need to check with SBE team , whether chip-op is allowed here. , anyway chi-op code is going to fail in this case and dump is triggerd , curently providing option to attempt chi-op.